client/driver/docker_progress.go

```diff
@@ -0,0 +1,242 @@
+	p.timestamp = time.Now()
+
+	lps := lpsFromString(msg.Status)
+	if lps == layerProgressStatusUnknown {
+		return
+	}
+
+	layer, ok := p.layers[msg.ID]
+	if !ok {
+		layer = &layerProgress{id: msg.ID}
+		p.layers[msg.ID] = layer
+	}
+	layer.status = lps
+	if msg.Progress != nil && lps == layerProgressStatusDownloading {
+		layer.currentBytes = msg.Progress.Current
+		layer.totalBytes = msg.Progress.Total
+	} else if lps == layerProgressStatusDownloaded {
+		layer.currentBytes = layer.totalBytes
+	}
+}
+
+func (p *imageProgress) currentBytes() int64 {
+	var b int64
+	for _, l := range p.layers {
+		b += l.currentBytes
+	}
+	return b
+}
+
+func (p *imageProgress) totalBytes() int64 {
+	var b int64
+	for _, l := range p.layers {
+		b += l.totalBytes
+	}
+	return b
+}
+
+type progressReporterFunc func(image string, msg string, timestamp time.Time, pullStart time.Time)
+
+type imageProgressManager struct {
+	*imageProgress
+	image            string
+	activityDeadline time.Duration
+	inactivityFunc   progressReporterFunc
+	reportInterval   time.Duration
+	reporter         progressReporterFunc
+	cancel           context.CancelFunc
+	stopCh           chan struct{}
+	buf              bytes.Buffer
+}
+
+func newImageProgressManager(
+	image string, cancel context.CancelFunc,
+	inactivityFunc, reporter progressReporterFunc) *imageProgressManager {
+	return &imageProgressManager{
+		image:            image,
+		activityDeadline: defaultPullActivityDeadline,
+		inactivityFunc:   inactivityFunc,
+		reportInterval:   defaultImageProgressReportInterval,
+		reporter:         reporter,
+		imageProgress: &imageProgress{
+			timestamp: time.Now(),
+			layers:    make(map[string]*layerProgress),
+		},
+		cancel: cancel,
+		stopCh: make(chan struct{}),
+	}
+}
+
+func (pm *imageProgressManager) withActivityDeadline(t time.Duration) *imageProgressManager {
+	pm.activityDeadline = t
+	return pm
+}
+
+func (pm *imageProgressManager) withReportInterval(t time.Duration) *imageProgressManager {
+	pm.reportInterval = t
+	return pm
+}
+
+func (pm *imageProgressManager) start() {
+	pm.pullStart = time.Now()
+	go func() {
+		ticker := time.NewTicker(defaultImageProgressReportInterval)
+		for {
+			select {
+			case <-ticker.C:
+				msg, timestamp := pm.get()
+				if time.Now().Sub(timestamp) > pm.activityDeadline {
+					pm.inactivityFunc(pm.image, msg, timestamp, pm.pullStart)
+					pm.cancel()
+					return
+				}
+				pm.reporter(pm.image, msg, timestamp, pm.pullStart)
+			case <-pm.stopCh:
+				return
+			}
+		}
+	}()
+}
+
+func (pm *imageProgressManager) stop() {
+	close(pm.stopCh)
+}
+
+func (pm *imageProgressManager) Write(p []byte) (n int, err error) {
+	n, err = pm.buf.Write(p)
+
+	for {
+		line, err := pm.buf.ReadBytes('\n')
+		if err == io.EOF {
+			pm.buf.Write(line)
+			break
+		}
+		if err != nil {
+			return n, err
+		}
+		var msg jsonmessage.JSONMessage
+		err = json.Unmarshal(line, &msg)
+		if err != nil {
+			return n, err
+		}
+
+		if msg.Error != nil {
```

Comment:
A comment as to what these errors mean would be useful.

Reply:
:heavy_check_mark: